connectors/sources/salesforce.py

```diff
@@ -0,0 +1,1475 @@
+                sobjects[record["Id"]] = record
+
+        return sobjects
+
+    async def _is_queryable(self, sobject):
+        """User settings can cause sobjects to be non-queryable
+        Querying these causes errors, so we try to filter those out in advance
+        """
+        return sobject.lower() in await self.queryable_sobjects()
+
+    async def _select_queryable_fields(self, sobject, fields):
+        """User settings can cause fields to be non-queryable
+        Querying these causes errors, so we try to filter those out in advance
+        """
+        sobject_fields = await self.queryable_sobject_fields()
+        queryable_fields = sobject_fields.get(sobject, [])
+        return [f for f in fields if f.lower() in queryable_fields]
+
+    def _parse_content_documents(self, record):
+        content_docs = []
+        content_links = record.get("ContentDocumentLinks", {}) or {}
+        content_links = content_links.get("records", []) or []
+        for content_link in content_links:
+            content_doc = content_link.get("ContentDocument")
+            if not content_doc:
+                continue
+
+            content_doc["linked_sobject_id"] = record.get("Id")
+            content_docs.append(content_doc)
+
+        return content_docs
+
+    def _combine_duplicate_content_docs(self, content_docs):
+        """Duplicate ContentDocuments may appear linked to multiple SObjects
+        Here we ensure that we don't download any duplicates while retaining links"""
+        grouped = {}
+
+        for content_doc in content_docs:
+            content_doc_id = content_doc["Id"]
+            if content_doc_id in grouped:
+                linked_id = content_doc["linked_sobject_id"]
+                linked_ids = grouped[content_doc_id]["linked_ids"]
+                if linked_id not in linked_ids:
+                    linked_ids.append(linked_id)
+            else:
+                grouped[content_doc_id] = content_doc
+                grouped[content_doc_id]["linked_ids"] = [
+                    content_doc["linked_sobject_id"]
+                ]
+                # the id is now in the list of linked_ids so we can delete it
+                del grouped[content_doc_id]["linked_sobject_id"]
+
+        grouped_objects = list(grouped.values())
+        return grouped_objects
+
+    async def _yield_non_bulk_query_pages(self, soql_query):
+        """loops through query response pages and yields lists of records"""
+        url = f"{self.base_url}{QUERY_ENDPOINT}"
+        params = {"q": soql_query}
+
+        while True:
+            response = await self._get_json(
+                url,
+                params=params,
+            )
+            yield response.get("records")
+            if response.get("done", True) is True:
+                break
+
+            url = f"{self.base_url}{response.get('nextRecordsUrl')}"
+            params = None
+
+    async def _execute_non_paginated_query(self, soql_query):
```

Comment:
I see this is used by `get_case_feeds` only. Why `case_feeds` doesn't need pagination?

Reply:
I was copying from the WS connector and thought (wrongly) that case_feeds were not being paginated (partly because we have a limit on the query of 500). I dug deeper and it _is_ paginated, so I'll fix that.

Reply:
case feeds also support pagination now